### PR TITLE
backtracking TM 

### PR DESCRIPTION
**Pull request by breznak**
merge new feature: Backtracking TM rewritten from Python to C++ by @dkeeney 

- [x] BackTM 
- [x] tests 
- [x] fix BackTM _checkStates() https://github.com/htm-community/nupic.cpp/pull/46#pullrequestreview-146512262 

Fixes: #45

### CODE REVIEW

**Comment by breznak** [comment]
thanks a lot @dkeeney ! I just pushed some minor fixes needed to get it compile on my side, please refresh before reviewing. 

Also note, due to all-code-changed diff in Segments, Cells4 & Array, I just took upstream version (may have missed some of yours additions, ie needed to add the getters in Cells4 too)

**Comment by breznak** [comment]
I figure the latest commit with test is https://github.com/chhenning/nupic.core/pull/4/commits/c745b6cfcdf8280b8353ec228babf25ba7d34391 , right? 

Unfortunately, the commit is changing too much stuff for me to merge/review, could you help me to reduce it to a bare minimum, please? 
Best would be if you can submit a commit to this branch that adds the tests (and other improvements to BackTM), or if you'd explain the minimum what I need? 
- the changes to BackTM
- BackTM test
- other classes with added functionality (Array, etc?) 
- the SP/TM-REgions are optional, right? I can still call the BackTM directly. Although I might add the classes later in a separate PR, please see my plans...

**Comment by dkeeney** [comment]
Well, that is going to be difficult to isolate the changes.
There has been a lot of work all over the code to
1) Made the unit tests run again (many had been commented out)
2) bring it up to C++17,
3) to add smart pointers, (including ArrayBase and subclasses)
4) replace boost:: with std::  (still a few to do)
5) fix serialization -- moving toward simple binary streaming,
                                Only a small part is using YAML and I am
thinking of replacing that.
6) Removed dimensions
7) Fixed Links so there is automatic type conversions or Zero Copy where
needed.

So this is not something that could be isolated to any specific module.
In general I tried to keep the documented API constant but internal calling
interfaces may have changed slightly in places. But to get everything in a
consistent set you need it all.  And you need to remove the Python modules.

My goals are:
- Remove ALL dependencies.  The last two are boost and YAML.
- C++ only library, in other words, be able to do the Hot Gym example
using C++ only, without any Python.
- Add the Python interface as a separate repository that links to the
nupic.cpp library.
- Add the CSharp interface as a separate repository that links to the
nupic.cpp library.

**Comment by dkeeney** [comment]
One more thing.  Git reports thousands of files being different. Most of
them are because we deleted all of the contents in the external folder (all
of the dependencies).  If you remove those first it helps but there are
still a lot of changes.

**Comment by breznak** [comment]
I totally align with your goals, but I want to merge them in per-partes, in a controlled manner, not just start anew all the files. For this PR I'm focusing only on the BackTM. 
Do you think the test would work only if I bring the backTM and tests from the last commit? Or any other files you'd recall with necessary changes? Thanks a bunch!

**Comment by dkeeney** [comment]
Well, I guess you could do that.  backTM by itself would not be very useful
without the TMRegion so it can be called from within the Network
Framework.  But if someone wanted to use it stand-alone it would probably
work.

It calls into Cells4 to do most of the work.  I had to add the
getters/setters in Cells4 but otherwise very little changed.

The backTM tried to not use any dependencies but I did use C++17 std::
functions for a few things. Note the use of the #include <any>  for the
values of a pair in a map (tmSavedState_t).  There is probably a better way
to do that.

For the unit tests...I had that CSVReader in there and I see I ended up not
using it so that #include can be removed.  Otherwise I think that should
run using gtest.

By the way, I have been using Visual Studio community 2017 as the build
environment.  Have not tried it with Linux environment yet.

**Comment by breznak** [comment]
> But if someone wanted to use it stand-alone it would probably work. 

That's what I want to achieve in this PR. The SP/TMRegions come next on plate. 

> It calls into Cells4 to do most of the work. I had to add the getters/setters in Cells4 but otherwise very little changed. 

check. I have already this version running in the CI here. 

> I did use C++17 std:: functions for a few things 

Will see what you added in the last commit. 

> CSVReader in there and I see I ended up not using it 

Thanks, good to know. Otherwise, I kinda liked this one https://github.com/ben-strasser/fast-cpp-csv-parser

> Have not tried it with Linux environment yet. 

Compiles fine here :+1:

**Comment by dkeeney** [comment]
In the interest of breaking up the changes I will see if I can isolate some other modules that could be merged one at a time.  Perhaps the ArrayBase and friends.

The problem I am having is that I cannot build the current Community/nupic.cpp under windows.  Some problem with a patch in 'Apr1StaticLib'.
1>Patching apr-1 sources in <SOURCE_DIR> via G:/AI/community/nupic.cpp/external/common/share/apr/apr.patch
1>'patch' is not recognized as an internal or external command,
1>operable program or batch file.

Perhaps it would be better to take some time and setup a Linux environment. I will need it to test with anyway.

**Comment by breznak** [comment]
Thanks, more modular/isolated patches would be awesome and make re-using much easier job for me.

> The problem I am having is that I cannot build the current Community/nupic.cpp under windows 

Let's resolve that. The community version is running a CI in Windows. 

> patch' is not recognized as an internal or external command, 

Did you try to install patch.exe?

>Perhaps it would be better to take some time and setup a Linux environment. I will need it to test with anyway. 

Or you can just try a PR to this repo and it'll run tests for all 3 platforms.

**Comment by dkeeney** [comment]
Apparently I do not have write permission so my approval does not very useful. So we need someone with write permission.

**Comment by breznak** [comment]
@dkeeney  added you to nupic.cpp devs, should work now. 

But actually, it's waiting for me further cherry-picking the tests. I'm doing it now, but it's a pain, next PR I'd like to do more of your changes, but I'll need help. I'm constantly afraid I miss some functionality from the merge.

**Comment by dkeeney** [comment]
Thanks.
I am testing with the new ArrayBase.cpp/hpp, Array.hpp, ArrayRef.hpp, and
BasicType.cpp/hpp. Should have it ready soon.

I did have to go to Ubuntu with VirtualBox.  I just did not want to clutter
up my Windows environment with a very old compiler just to build for Python
2.7.

Once we move the Python interface to a separate repository I can go back to
building on Windows.

**Comment by breznak** [comment]
@dkeeney could you please help me cherry-pick only the BackTM tests to this PR, so I can merge it? I got stuck on that with some errors somehow. 

Today I'd like to merge this PR and ArrayBase. Then move to c++17, and remove serialization afterwards during weekend.

**Comment by dkeeney** [comment]
backtrackingTMCppTests.cpp  is the test you want.  The tests for serialization could be commented out until we get the serialization switched over everywhere.

**Comment by dkeeney** [comment]
@breznak, What would you like me to work on next?  I don't want to conflict
with what you are doing.

   - I could work on the Network framework (the stuff in the engine
   directory).
   - or, something smaller, I could replace the Yaml serialization in
   Network with streaming version.  I coded that once and I can pull it out
   again.
   - perhaps the rest of the routines in the ntypes and types directories.
   Not may more changes there.  We did get rid of the NTA types in Types.h
   (NTA_Byte, NTA_SIZE, NTA_UInt16, etc.)

**Comment by breznak** [comment]
Many thanks @dkeeney ! Actually, if you could help me first resolving the tests here? The BackTM itself compiled ok, but I'm stuck on  c503d4 where the newest changes are merged + addst tests; 

I had to "revert" some of your changes (going back to boost:: instead of c++17 std::) etc, but there are still some problems and I', stuck.

**Comment by dkeeney** [comment]
@breznak I am looking at BackTM.  You changed the outputType from char[] to
string.  That is fine, I agree, but the loc structure was setup to be
written with a binary stream during serialization...so it cannot contain a
string.

I will try to fix it up for you.

**Comment by dkeeney** [comment]
@breznak  I made a few changes and got it all to build and pass the unit
tests.
However I did have to comment out the test Cells4Test, testEqualsOperator.
The problem is that Cells4 uses a random number generator that is a
singleton. So if you create two Cells4 objects they are guaranteed to be
different even if initialized with the same parameters.  The only thing
that the equals operator can be used for is when comparing a serialized
Cells4 object against its the restored Cells4 object because the Random
object state is also serialized.

You might be able to get the testEqualsOperator test to work by messing
with the Random object's states such that they are the same state at the
beginning of each operation that a Cells4 object makes.  However, its more
trouble than it is worth.  The backtrackingTM tests does compare serialized
Cells4 objects.  The TMRegion tests do a much more extensive test because
there the Random objects are different (different Network objects).

Anyway I pushed my changes back into your feature_dkeeny_backtrackingTM
branch where you can check it out.

Dave

File: src/nupic/algorithms/BacktrackingTMCpp.cpp [hunk 1]
```diff
@@ -89,7 +89,8 @@ BacktrackingTMCpp::BacktrackingTMCpp(
     Int32 verbosity, bool checkSynapseConsistency, UInt32 pamLength,
     UInt32 maxInfBacktrack, UInt32 maxLrnBacktrack, UInt32 maxAge,
     UInt32 maxSeqLength, Int32 maxSegmentsPerCell, Int32 maxSynapsesPerSegment,
-    const std::string outputType) {
+    std::string outputType)
```
Comment (breznak) [inline src/nupic/algorithms/BacktrackingTMCpp.cpp]: nit, make this `const` if possible

File: src/nupic/algorithms/BacktrackingTMCpp.cpp [hunk 2]
```diff
@@ -802,72 +804,85 @@ std::shared_ptr<Real> BacktrackingTMCpp::getConfHistogram() {
...
-  ss[name] = target; // 'target' could be a shared_ptr<Byte> or shared_ptr<Real>
+                                      Byte *buf, Size count) {
+  std::shared_ptr<Byte> target(new Byte[count], std::default_delete<Byte[]>());
+  fastbuffercopy<Byte>(target.get(), buf, count);
+  struct ss_t val;
+  val.Byteptr = target;
+  ss[name] = val;
+}
+void BacktrackingTMCpp::deepcopySave_(tmSavedState_t &ss, std::string name,
+                                      Real *buf, Size count) {
```
Comment (breznak) [inline src/nupic/algorithms/BacktrackingTMCpp.cpp]: keep the templated method? Or make a private templated; and overload with "public"/protected Byte/Real variants?

File: src/nupic/algorithms/BacktrackingTMCpp.cpp [hunk 3]
```diff
@@ -1810,10 +1826,11 @@ void BacktrackingTMCpp::load(std::istream &in) {
   NTA_ASSERT(len == sizeof(loc_));
   in.ignore(1);
   in.read((char *)&loc_, len);
+  in >> outputType_;
 
-  NTA_ASSERT(loc_.numberOfCols = cells4_->nColumns());
-  NTA_ASSERT(loc_.cellsPerColumn = cells4_->nCellsPerCol());
-    nCells = cells4_->nCells();
+  NTA_ASSERT(loc_.numberOfCols == cells4_->nColumns());
```
Comment (breznak) [inline src/nupic/algorithms/BacktrackingTMCpp.cpp]: nice catch!

File: src/nupic/algorithms/Cells4.cpp [hunk 4]
```diff
@@ -30,7 +30,7 @@
 #include <sstream>
 #include <vector>
 
-#include <boost/filesystem.hpp>
+//#include <boost/filesystem.hpp>
```
Comment (breznak) [inline src/nupic/algorithms/Cells4.cpp]: nit, we can remove the inclde

File: src/nupic/algorithms/Cells4.cpp [hunk 5]
```diff
@@ -67,18 +67,39 @@ static nupic::Timer chooseCellsTimer;
...
+                     Real permConnected,
+                     Real permMax,
+                     Real permDec,
+                     Real permInc,
+                     Real globalDecay,
+                     bool doPooling,
+                     int seed,
+                     bool initFromCpp,
+                     bool checkSynapseConsistency)
+  : _rng(seed < 0 ? rand() : seed)
```
Comment (breznak) [inline src/nupic/algorithms/Cells4.cpp]: unnecessary change for this PR? May it break the other 3 tests? 
// I gues it complies with the API?

File: src/nupic/algorithms/Cells4.cpp [hunk 6]
```diff
@@ -1999,16 +2157,15 @@ void Cells4::save(std::ostream &outStream) const {
...
 
 //----------------------------------------------------------------------------
 /**
  * Save the state to the given file  (must be binary file).
  */
 void Cells4::saveToFile(std::string filePath) const {
-  boost::filesystem::path p(filePath.c_str());
-  boost::filesystem::create_directories(p.parent_path());
+//do later  boost::filesystem::path p(filePath.c_str());
+//do later  boost::filesystem::create_directories(p.parent_path());
```
Comment (breznak) [inline src/nupic/algorithms/Cells4.cpp]: ok, i take back the "remove include <filesystem>

File: src/nupic/algorithms/Cells4.cpp [hunk 7]
```diff
@@ -2994,3 +3151,66 @@ void Cells4::resetTimers() {
   chooseCellsTimer.reset();
 #endif
 }
+
+/**
+* Compare against another Cells4.
```
Comment (breznak) [inline src/nupic/algorithms/Cells4.cpp]: add the comment about limitation of equals?

File: src/nupic/algorithms/SegmentUpdate.cpp [hunk 8]
```diff
@@ -72,13 +72,28 @@ bool SegmentUpdate::invariants(Cells4 *cells) const {
...
-  if (_cellIdx != o._cellIdx || _segIdx != o._segIdx ||
-      _sequenceSegment != o._sequenceSegment || _timeStamp != o._timeStamp ||
-      _phase1Flag != o._phase1Flag ||
-      _weaklyPredicting != o._weaklyPredicting) {
-    return false;
+  // synapses
+  if (s._synapses.size() != _synapses.size()) return false;
+  for (Size synIdx1 = 1; synIdx1 < _synapses.size(); synIdx1++)
+  {
+      if (s._synapses[synIdx1] != _synapses[synIdx1])
```
Comment (breznak) [inline src/nupic/algorithms/SegmentUpdate.cpp]: extra check of elements in the for. Is this an enhancement? Or in the prev version the Random ensured same values so they could skip this?

File: src/nupic/algorithms/SegmentUpdate.hpp [hunk 9]
```diff
@@ -191,16 +186,22 @@ class SegmentUpdate : Serializable<SegmentUpdateProto> {
...
 
-//--------------------------------------------------------------------------------
-#ifndef SWIG
-inline std::ostream &operator<<(std::ostream &outStream,
-                                const SegmentUpdate &update) {
-  update.print(outStream);
-  return outStream;
-}
-#endif
+   bool operator==(const SegmentUpdate &s) const { return this->equals(s); }
```
Comment (breznak) [inline src/nupic/algorithms/SegmentUpdate.hpp]: why do you use separate `equals()` and not overload `==` directly?

File: src/test/unit/algorithms/BacktrackingTMCppTest.cpp [hunk 10]
```diff
@@ -29,15 +29,19 @@
  *
  * For those not familiar with GTest:
  * ASSERT_TRUE(value)   -- Fatal assertion that the value is true. Test
- *terminates if false. ASSERT_FALSE(value)   -- Fatal assertion that the value
```
Comment (breznak) [inline src/test/unit/algorithms/BacktrackingTMCppTest.cpp]: maybe remove these? GTest is used everywhere here, people are either familiar, or can find the meanings?

File: src/test/unit/algorithms/Cells4Test.cpp [hunk 11]
```diff
@@ -593,6 +593,8 @@ TEST(Cells4Test,
 /**
  * Test operator '=='
  */
+/******** This cannot work because of the Random number generator.
+
```
Comment (breznak) [inline src/test/unit/algorithms/Cells4Test.cpp]: TODO myself, investigate if safe assumption

**Comment by breznak** [comment]
> You might be able to get the testEqualsOperator test to work by messing with the Random object's states such that they are the same state at the beginning of each operation that a Cells4 object makes. However, its more trouble than it is worth. ... 

Agree

> ... The problem is that Cells4 uses a random number generator that is a singleton. So if you create two Cells4 objects they are guaranteed to be different even if initialized with the same parameters. 

Is this what we want? I'd say if you have same params, esp. same `seed`, you'll get replicable numbers. If you wanted 2 objects with different rng, use different seed. 

> The only thing that the equals operator can be used for is when comparing a serialized Cells4 object against its the restored Cells4 object because the Random object state is also serialized. 

In your implementation this makes sense. I'll need to understand more how they made it work before and what was it for? There was the `memcpy` in Random.equals for that?

**Comment by dkeeney** [comment]
>
> nit, make this const if possible
>
an std::string is immutable so it is already effectively a const.

keep the templated method? Or make a private templated; and overload with
> "public"/protected Byte/Real variants?
>
I was just trying to get rid of the std::any.   This was the first thing I
thought of.  There are probably better ways but from this it is obvious
that the values are Byte or Real.

unnecessary change for this PR? May it break the other 3 tests?
>
This was how the original code had it...I just happened to stumble onto
this.  You are right, this has nothing to do with the PR.

add the comment about limitation of equals?
>
I did comment about it someplace.   But it makes since to do it here also.

extra check of elements in the for. Is this an enhancement? Or in the prev
> version the Random ensured same values so they could skip this?
>
I was not sure what might have ensured the same values.  You can remove it
if you are sure that will always be equal.

why do you use separate equals() and not overload == directly?
>
Just force of habit.  spent too much time in Java :)

maybe remove these? GTest is used everywhere here, people are either
> familiar, or can find the meanings?
>
Actually I added that when I started working on this because I was not
familiar enough with gtest to remember them.  I have used them enough now
that I don't need the crutch but perhaps someone else new might find it
useful.

TODO myself, investigate if safe assumption
>
Yes, please do. I think there is a way to capture the random function such
that there can be separate Randoms for each of the Cells4.  Don't know how
to do that however. Perhaps with some research you could figure it out.

I will assume you will finish that branch.  I am off working or removing
capnproto.  I will try to keep focused on just the PR and avoid changing
other things I happen to run into :)

Dave
=====

File: src/nupic/algorithms/BacktrackingTMCpp.cpp [hunk 12]
```diff
@@ -0,0 +1,1863 @@
...
+  return (numActiveSyns >= threshold);
+}
+
+/**
+  struct SegOnCellInfo_t {
+      Size c;
+      Size i;
+      Size segIdx;
+      bool isSequenceSegment;
+      Size positi
```
Comment (breznak) [inline src/nupic/algorithms/BacktrackingTMCpp.cpp]: remove for now, will be fixed in #57

File: src/nupic/algorithms/BacktrackingTMCpp.hpp [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+ *            - 'normal': output the OR of the active and predicted state.
+ *            - 'activeState': output only the active state.
+ *            - 'activeState1CellPerCol': output only the active
+ *               state, and at most 1 cell/column. If more than 1 cell
+ *               is active in a column, the one with the highest confidence
+ *               is sent up.
+ *
+ *************/
+
+class BacktrackingTMCpp {
```
Comment (breznak) [inline src/nupic/algorithms/BacktrackingTMCpp.hpp]: should extend TempralMemory 
so we can use
`TemporalMemory tm = (Backtracking)TemporalMemory(...)`;

File: src/nupic/algorithms/BacktrackingTMCpp.hpp [hunk 14]
```diff
@@ -0,0 +1,586 @@
...
+  void setRetrieveLearningStates(bool val) { loc_.retrieveLearningStates = val; }
+
+  /**
+   * Serialization. save and load the current state of the TM to/from the
+   * specified file.
+   */
+  void saveToFile(std::string filePath);
+  void loadFromFile(std::string filePath);
+  void save(std::ostream &out) const;
+  void load(std::istream &in);
```
Comment (breznak) [inline src/nupic/algorithms/BacktrackingTMCpp.hpp]: TODO remove for now

File: src/nupic/algorithms/Cells4.cpp [hunk 15]
```diff
@@ -2115,202 +2076,232 @@ void Cells4::save(std::ostream &outStream) const {
...
-  outStream << _learnActiveStateT << " " << _learnActiveStateT1 << " "
-            << _learnPredictedStateT << " " << _learnPredictedStateT1
-            << std::endl;
-
-  // Additions in version 2.
-  outStream << _segmentUpdates.size() << " ";
+
+  // Capture the class variables and write them as a binary block.
+  struct self_t self;
+  memset(&self, 0, sizeof(self));
```
Comment (breznak) [inline src/nupic/algorithms/Cells4.cpp]: check if the memset failed?

File: src/nupic/algorithms/Cells4.hpp [hunk 16]
```diff
@@ -60,1092 +61,1182 @@
...
-  void updateSegment(UInt colIdx, UInt cellIdxInCol, UInt segIdx,
-                     const std::vector<std::pair<UInt, UInt>> &extSynapses);
-
-  //-----------------------------------------------------------------------
-  /**
-   * Rebalances and rebuilds internal structures for faster computing
-   *
-   */
-  void _rebalance();
-
```
Comment (breznak) [inline src/nupic/algorithms/Cells4.hpp]: David, please figure how your IDE (auto)transforms code. I think there is a code-style for c++ nupic.

**Comment by dkeeney** [comment]
>
> *typo dense_buffer *
>
good catch.

*should be moved to #13  (nonzero method)*
>
There is probably already a method that does this in the utilities
someplace but I could not find one without doing a lot more looking so I
just made one.  So if you can find one, use it.

*TODO: use printVector from VectorHelpers *
>
Perhaps...just don't change the function signature. It is part of the API.

*remove for now, will be fixed in #57
> <https://github.com/htm-community/nupic.cpp/pull/57>*
>
do not understand this one...but if it is related to capnproto it will be
removed.

*should extend TempralMemory....*
>
We cannot change the signature of TempralMemory...Python is expecting it
the way it is.  There is another component,  "temporal_memory_shim.py",
that I have yet to port to C++ which wraps BacktrackingTM with the calling
signature for temporal_memory. And another one, "backtracking_tm_shim.py",
that wraps temporal_memory with the signature for BacktrackingTM.  This
second one needs to be called by TMRegion so that both backtrackingTM and
temporal_memory can be accessed as part of the Network framework if not
using any Python.  So, yes this is in the plan.

* TODO remove for now *
>
Do not understand. That is all needed.

*check if the memset failed? *
>
If memset fails there will be an Access Violation exception and everything
dies. They will know

*code-style for c++ nupic. *
>
http://nupic.docs.numenta.org/stable/contributing/cpp-style-guide.html
Unfortunately they do not mention a clangFormat style to use and they do
not indicate placement of  {.
I am using the google style clangFormat with tabs set at 2. This should
wrap at 100 characters but it still seems to wrap at 80 which is really
annoying.  I don't know what @chhenning used but I suspect it was the
Visual Studio default.

The existing code does not seem to be consistent. It was obviously written
by many different people with different styles. I propose that we run it
all through a clang formatter and then work from there.  The function
header comments all need to be cleaned up if they are to be used by
doxygen...perhaps a project to do after we have it working so we know what
functions are remaining. In-code documentation in general is quite poor.  I
would like to see design discussions in the code so it is obvious what the
intent was but some people think that should be someplace else.

By the way, I am now using Beyond compare as a comparison tool. It does a
good job and works on both linux and windows. whitespace differences are in
blue but real differences are in red so they are easy to spot.  And it can
be used as the editor without formatting :)

**Comment by breznak** [comment]
>> *remove for now, will be fixed in #57 

> do not understand this one...but if it is related to capnproto it will be
removed.

My bad. I'm trying to minimize this PR as still trying to resolve the errors and Random. I'll leave the (new) serialization in place. 

> There is another component, "temporal_memory_shim.py", that I have yet to port to C++ which wraps BacktrackingTM with the calling signature for temporal_memory. And another one, "backtracking_tm_shim.py", that wraps temporal_memory with the signature for BacktrackingTM. 

This sounds too complicated and legacy burden. Wouldn't it be easier then providing new TemporalMemoryInterface(Abstract) and extending TM and BackTM from that then? Region could depend on that class. Or some other, cleaner solution? 

> Unfortunately they do not mention a clangFormat style to use and they do not indicate placement of {. 

There is/used to be a lint format for nupic.core. I recall esp this change of `{` (at the same line). 
I guess easies for now would be using git ignore whitespace 
https://medium.com/therobinkim/git-add-everything-but-whitespace-changes-deec3dce39f 
so each of us can format whichever way (doing "reformat" is also a way, although it would touch all files and break git blame, also any diffs to other repos would be crazy, )

> By the way, I am now using Beyond compare as a comparison tool. It does a good job and works on both linux 

I'll check it out, thanks! :+1:

**Comment by breznak** [comment]
@dkeeney I have a tricky question for you ;) 

Status of this PR: 
- [ ] py.test bindings/py/tests/ fail due to problems with capnp -> won'tfix here, as will be done in your PR. 
- [ ] tricky: random failures of BackTMTest.testCheckpointMiddleOfSequence :crab: 

> make -j4 && make install && make tests_all 

```
Done -- all tests passed
Memory usage: 100 (virtual) 100 (real) at iteration 0
--- ALL TESTS PASSED ---
[100%] Built target tests_cpp_region
[100%] Running all tests
```

> ../release/bin/unit_tests --gtest_filter="Back*Middle*" 

```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from BacktrackingTMTest
[ RUN      ] BacktrackingTMTest.testCheckpointMiddleOfSequence
[       OK ] BacktrackingTMTest.testCheckpointMiddleOfSequence (48 ms)
[----------] 1 test from BacktrackingTMTest (48 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (48 ms total)
[  PASSED  ] 1 test.
```

BUT!
>  ../release/bin/unit_tests --gtest_filter="Back*Check*"

```
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from BacktrackingTMTest
[ RUN      ] BacktrackingTMTest.testCheckpointLearned
TM's match
[       OK ] BacktrackingTMTest.testCheckpointLearned (53 ms)
[ RUN      ] BacktrackingTMTest.testCheckpointMiddleOfSequence
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/test/unit/algorithms/BacktrackingTMCppTest.cpp:337: Failure
Value of: tm2
  Actual: 224-byte object <58-53 4C-F2 E3-55 00-00 64-00 00-00 0C-00 00-00 02-00 00-00 00-7F 00-00 2A-00 00-00 C7-E2 C5-59 00-00 00-00 00-00 00-00 32-00 00-00 00-00 00-00 00-00 00-00 01-00 00-00 00-00 6C-F3 00-00 80-BF ... 40-FD A6-F3 E3-55 00-00 00-B4 7D-F3 E3-55 00-00 0F-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 90-13 73-F3 E3-55 00-00 10-BF 7D-F3 E3-55 00-00 10-90 78-F3 E3-55 00-00>
Expected: tm1
Which is: 224-byte object <58-53 4C-F2 E3-55 00-00 64-00 00-00 0C-00 00-00 02-00 00-00 00-7F 00-00 2A-00 00-00 C7-E2 C5-59 00-00 00-00 00-00 00-00 32-00 00-00 00-00 00-00 00-00 00-00 01-00 00-00 00-00 6C-F3 00-00 80-BF ... 90-D2 70-F3 E3-55 00-00 60-6D 76-F3 E3-55 00-00 0F-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 40-9F 72-F3 E3-55 00-00 00-90 70-F3 E3-55 00-00 60-79 79-F3 E3-55 00-00>
Both TM trained
[  FAILED  ] BacktrackingTMTest.testCheckpointMiddleOfSequence (69 ms)
[----------] 2 tests from BacktrackingTMTest (122 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (122 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] BacktrackingTMTest.testCheckpointMiddleOfSequence
```

I'm suspecting: 
- bug in the test itself? 
- some pointer black-magic in Random/RandomImpl ? 
- pointer magic in BackTM? (we don't use smart pointers here, yet) 

Nice, huh?

File: src/nupic/algorithms/BacktrackingTMCpp.cpp [hunk 17]
```diff
@@ -560,16 +560,22 @@ std::shared_ptr<struct BacktrackingTMCpp::predictionResults_t> BacktrackingTMCpp
...
+    for (UInt i=0; i < pattern.size(); i++) {
+	    //debug
+	    std::cout << "=====================================\n";
+	   cout << "pattern= ";
+	   for (auto e : pattern) cout << e << ", ";
+	  cout << "confidences= ";
+	  for(UInt i=0; i < sizeof(colConfidence)/sizeof(colConfidence[0]); i++) cout << colConfidence[i] << ", ";
+//      NTA_ASSERT(pattern[i] < positiveColumnCount) << "BackTM: " << pattern[i] << " of " << positiveColumnCount;
+      //end-debug
+      positivePredictionSum += colConfidence[i];
```
Comment (breznak) [inline src/nupic/algorithms/BacktrackingTMCpp.cpp]: - colConfidences are always 0 ? (but because first run?)
- this crashed, because addressing colConfidence with `pattern[i]` results in out of bounds exception. But both patterm and colConfidence have same size. What is the correct logic to compute positivePredictionSum?

**Comment by dkeeney** [comment]
The real logic is in
https://github.com/numenta/nupic/blob/master/src/nupic/algorithms/backtracking_tm.py
line 2928.
I looked at the logic and it looks fine to me...but then I am probably not
seeing the mistake because this is how I interpreted the Python.
If the colConfidences array is empty then it may be getting a bad array
from Cells4 colConfidenceT

**Comment by breznak** [comment]
Please review (although it's your code, so I'll review too) 
I've updated to current code, fixed some issues, and it should be now working!

**Comment by dkeeney** [comment]
BacktrackingTMCppTest.cpp:42:13: error: unused variable 'verbose' .
Just delete the declaration for verbose.  It was for my debugging.

**Comment by dkeeney** [comment]
I did this port on MSVC without SWIG.  I would be surprised if there were NOT problems when compiling with gcc.

**Comment by breznak** [comment]
I’ve fixed the problems on Linux, compiles and runs tests ok here

I’ll fix the unused variable

-- 
Marek Otahal :o)

**Comment by breznak** [comment]
@dkeeney @ctrl-z-9000-times please review if you have time, I'd like to merge before the big pybind changes

File: src/nupic/algorithms/Cells4.cpp [hunk 18]
```diff
@@ -30,7 +30,6 @@
 #include <sstream>
 #include <vector>
 
-#include "cycle_counter.hpp"
```
Comment (dkeeney) [inline src/nupic/algorithms/Cells4.cpp]: We no longer need cycle_counter.hpp?  Cool.  We need to delete that from the dependencies.

File: src/nupic/algorithms/Cells4.cpp (hunk 5 above)
Comment (dkeeney) [reply src/nupic/algorithms/Cells4.cpp]: It looks so much more readable.

File: src/nupic/algorithms/Cells4.cpp (hunk 18 above)
Comment (breznak) [reply src/nupic/algorithms/Cells4.cpp]: good spot!

**Comment by breznak** [comment]
Thanks for your contribution @dkeeney , BackTM is finally part of the nupic cpp! :clap:
